[Workspace][Feature] Add ACL related functions

gaobinlong opened the pull request:
### Description
The main change of this PR are:

- Add `permissions` field to the default mapping of the OSD index(.kibana) to support ACL for workspace
- Update the test snapshot after we change the default mapping of the OSD index
- Add ACL class and basic functions

### Issues Resolved
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5083

## Testing the changes

<!--
  Please provide detailed steps for validating your changes. This could involve specific commands to run,
  pages to visit, scenarios to try or any other information that would help reviewers verify
  the functionality of your change
-->

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
  - [x] `yarn test:ftr`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

xluo-aws commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,249 @@
+   *     }
+   * }
+   *
+   * transformed permissions: [
```
Could you add why we need to transform the format?

gaobinlong replied:
Sure, I'll add that, thanks!

Hailong-am commented on src/core/server/saved_objects/permission_control/acl.test.ts:
```diff
@@ -0,0 +1,261 @@
+
+import { Principals, Permissions, ACL } from './acl';
+
+describe('SavedObjectTypeRegistry', () => {
```
should we have a name more related to ACL instead of `SavedObjectTypeRegistry`?

gaobinlong replied:
Good catch, thanks!

Hailong-am commented on src/core/server/saved_objects/permission_control/acl.test.ts:
```diff
@@ -0,0 +1,261 @@
+            {
+              terms: {
+                type: ['workspace', 'index-pattern'],
+              },
```
how about add another case to cover type is undefine?

xluo-aws commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,249 @@
+        principals.users,
+        principals.groups
+      );
+      if (result) {
```
Is this needed? permissions[permissionType] is already updated when calling deleteFromPrincipals

gaobinlong replied:
By my test, in this case, the function call only passes value of `this.permissions[permissionType]` , not reference, so the original value will not change if we do not update it by the function call result.

xluo-aws replied:
I c. You can mark it as resolved.

xluo-aws commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    for (const permissionType of permissionTypes) {
+      this.permissions[permissionType] = addToPrincipals(
```
do we need to assign the result back if it's already updated in addToPrincipals?

xluo-aws commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,249 @@
+    if (!principals || !permissionTypes) {
+      return {
+        query: {
+          match_none: {},
```
If there is no principals but has permissionTypes, should we return all?

gaobinlong replied:
I think it's better that the caller skips the function call if there is no principal, this judge is just a parameter check which means no principal shouldn't happen, and for security reason, we should return none here if no principal.

kavilla commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,250 @@
+
+    permissionTypes.forEach((permissionType) => {
+      Object.entries(principals).forEach(([principalType, principalsInCurrentType]) => {
+        subBool.should.push({
```
@joshuarrrr @ashwin-pc do you guys know if this stuff can be built used by the data plugin?

zengyan-amazon replied:
not likely in data plugin, since data plugin deals with index data in OpenSearch, while this ACL is for saved objects.

kavilla reviewed (commented):
Thanks for opening!

IMO, it might be worth to have this one hooked into the saved object functionality for us to see how it is being used and to give suggestions.

I see it on the mappings but I'm not sure how it interacts for existing data barring the mapping addition.

Also, just a heads up but I'm sure you already know.  I'm really easily able to pass a compressed file of a 100+ MB files to the saved objects API. But the default size of the saved objects payload is quite large. Some users saved objects are sometimes too large that it is really not as performant. So based off the design proposal seeing how these permission controls are being used might make a big difference for our suggestions and increased latency to a potential slower response.

Would also suggest adding a configuration [here](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/saved_objects_config.ts#L48) that gives cluster admins the ability to disable permission controls? But I'm not sure how that impacts your design proposal but would make me sleep easier at night when we are filtering responses.

Final thing, which I tried seeing in the [design](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4633/files?short_path=92b5768#diff-92b5768e5935d8ba3bc3f4fb9058e364330aa9628ae1ca25e2a7c1dd3139fbd0), I'm pretty sure the saved objects retrieval is slightly cached and it might refresh at specific times. So I can see  it having an impact on design if we want to worry about where to store this information but seems like a difficult issue to solve. For example, what do we plan to do as a user who originally had access, could still get the saved objects. Will they have the ability to have write permissions still in the cache state. They also still have access to the data and resending the request without refreshing the page.

I will paste a links to this on the design doc.

Thanks again!

abbyhu2000 commented on src/core/server/saved_objects/permission_control/acl.test.ts:
```diff
@@ -0,0 +1,348 @@
+      .removePermission(['write'], {
+        groups: ['group2'],
+      })
+      .removePermission(['write'], {
```
one question about `removePermission`:

From the documentation, it seems like write permission includes the read permission. 
For example, if a user have both read and write permission to the saved object, and we call `removePermission` to remove the read permission. In this case, is the write permission still valid?

Or should we set a rule that we can not remove read permission while there is still a valid write permission?

abbyhu2000 commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,250 @@
+  }
+
+  // permissions object build function, remove specific permission of specific principal from the object
+  public removePermission(permissionTypes: string[], principals: Principals) {
```
one question about `removePermission`:

From the documentation, it seems like write permission includes the read permission. 
For example, if a user have both read and write permission to the saved object, and we call `removePermission` to remove the read permission. In this case, is the write permission still valid?

Or should we set a rule that we can not remove read permission while there is still a valid write permission? Or if the rules have been set, then can we add some comments to explain it?

xluo-aws replied:
For acl permission, it's either read or write, user won't have both permissions.  It's the workspace saved object permission check function that has this logic. For example,  if a user need read permission on a saved object, saved object permission will call ACL hasPermission function to check whether user has either read or write permission on that saved object. removePermission just remove the specified permission. We'll add comment to make it clear.

xluo-aws replied:
For acl permission, it's either read or write, user won't have both permissions.  It's the workspace saved object permission check function that has this logic. For example,  if a user need read permission on a saved object, saved object permission will call ACL hasPermission function to check whether user has either read or write permission on that saved object. removePermission just remove the specified permission.

kavilla replied:
@xluo-aws would you be able to add some inline comments like the structure defined here: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5075#discussion_r1339606507

xluo-aws replied:
Sure

abbyhu2000 commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,250 @@
+  return principals;
+};
+
+const checkPermission = (currentPrincipals: Principals | undefined, principals: Principals) => {
```
Just to verify: Is the user's permission always be prioritized before the group's permission? 
For example, if a user and the user's group has been given different type of permission, and the group have higher permission than the user, what will the user's permission be?

xluo-aws replied:
We only check if user or the group they are associated with have the specified permission or not, so there is no difference whether the permission is from the user or associated group.

abbyhu2000 commented:
Thanks for the detailed testing, but I think the logics can be better defined and documented on some of the helper functions of the permissions.

xluo-aws commented:
> Thanks for opening!
> 
> IMO, it might be worth to have this one hooked into the saved object functionality for us to see how it is being used and to give suggestions.
> 
> I see it on the mappings but I'm not sure how it interacts for existing data barring the mapping addition.
> 

@kavilla, Rocky,  the code that calls ACL functions will be submitted in another PR so it has to wait for this PR to be merged first. You can either reference https://github.com/ruanyl/OpenSearch-Dashboards/tree/workspace or we can create a draft PR if broken build is acceptable. What's your preference?

xluo-aws commented:
> Also, just a heads up but I'm sure you already know. I'm really easily able to pass a compressed file of a 100+ MB files to the saved objects API. But the default size of the saved objects payload is quite large. Some users saved objects are sometimes too large that it is really not as performant. So based off the design proposal seeing how these permission controls are being used might make a big difference for our suggestions and increased latency to a potential slower response.

Thanks for bring this up, we are not aware of there could be such big saved objects. Is that from one of the customer plugin? For most case, we should be fine because we just check the workspace's ACL and workspace document is small; however if user want to access a specified saved object, we'll get the workspace of that saved object first then check that workspace's ACL to determine if you user has necessary permission. If user has permission, we'll read the object again to fetch the required fields. So permission check will add one more call. To optimize it, we can consider reading the docvalue of workspace field so the whole saved object won't be loaded. Will that work? If so, will need to update repository to add docvalue field as an optional parameter?

xluo-aws commented:
> Final thing, which I tried seeing in the [design](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4633/files?short_path=92b5768#diff-92b5768e5935d8ba3bc3f4fb9058e364330aa9628ae1ca25e2a7c1dd3139fbd0), I'm pretty sure the saved objects retrieval is slightly cached and it might refresh at specific times. So I can see it having an impact on design if we want to worry about where to store this information but seems like a difficult issue to solve. For example, what do we plan to do as a user who originally had access, could still get the saved objects. Will they have the ability to have write permissions still in the cache state. They also still have access to the data and resending the request without refreshing the page.
>

@kavilla Rocky, could you clarify which cache you are talking about? We don't explicitly cache any data and the permission evaluation is done at real time when user submit the CRUD request.

xluo-aws commented:
> 
> Would also suggest adding a configuration [here](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/saved_objects_config.ts#L48) that gives cluster admins the ability to disable permission controls? But I'm not sure how that impacts your design proposal but would make me sleep easier at night when we are filtering responses.
> 

Good suggestion, we'll think about if we want to disable the ACL check statically or dynamically.

kavilla commented:
> > Thanks for opening!
> > IMO, it might be worth to have this one hooked into the saved object functionality for us to see how it is being used and to give suggestions.
> > I see it on the mappings but I'm not sure how it interacts for existing data barring the mapping addition.
> 
> @kavilla, Rocky, the code that calls ACL functions will be submitted in another PR so it has to wait for this PR to be merged first. You can either reference https://github.com/ruanyl/OpenSearch-Dashboards/tree/workspace or we can create a draft PR if broken build is acceptable. What's your preference?

I think a draft on the usage will be helpful. Or a set a files to look at within that branch.

Thank you.

kavilla commented:
> > Would also suggest adding a configuration [here](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/saved_objects_config.ts#L48) that gives cluster admins the ability to disable permission controls? But I'm not sure how that impacts your design proposal but would make me sleep easier at night when we are filtering responses.
> 
> Good suggestion, we'll think about if we want to disable the ACL check statically or dynamically.

Sounds good, do y'all have a board or a meta issue that is tracking things that will be come back to before finalizing ?

kavilla commented:
> > Final thing, which I tried seeing in the [design](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4633/files?short_path=92b5768#diff-92b5768e5935d8ba3bc3f4fb9058e364330aa9628ae1ca25e2a7c1dd3139fbd0), I'm pretty sure the saved objects retrieval is slightly cached and it might refresh at specific times. So I can see it having an impact on design if we want to worry about where to store this information but seems like a difficult issue to solve. For example, what do we plan to do as a user who originally had access, could still get the saved objects. Will they have the ability to have write permissions still in the cache state. They also still have access to the data and resending the request without refreshing the page.
> 
> @kavilla Rocky, could you clarify which cache you are talking about? We don't explicitly cache any data and the permission evaluation is done at real time when user submit the CRUD request.

So the saved objects are not refreshed every time a user is interacting with them. 

Say the application is hosted on a cloud, and a user navigates to the dashboard page, their saved objects are loaded with their permissions based on what is stored within the engine (in this case OpenSearch).

Even if the user submits a request for new data or adds a new filter it is not refreshing it's saved objects data. Could it be possible for some admin of a workspace updated the actual information stored within `.kibana`. Which means dashboards and visualizations can be still populating with data to stuff they no longer have permissions until they refresh the page or navigate to a part of the application that update the cached saved objects.

kavilla commented:
> > Also, just a heads up but I'm sure you already know. I'm really easily able to pass a compressed file of a 100+ MB files to the saved objects API. But the default size of the saved objects payload is quite large. Some users saved objects are sometimes too large that it is really not as performant. So based off the design proposal seeing how these permission controls are being used might make a big difference for our suggestions and increased latency to a potential slower response.
> 
> Thanks for bring this up, we are not aware of there could be such big saved objects. Is that from one of the customer plugin? 

>  Is that from one of the customer plugin?

Nah, since these saved objects and workspaces are public APIs, people will write their own scripts to ran by their pipelines to hit these functions. Usually as a part of a lifecycle thing to generate a new instance of OpenSearch Dashboards with the same data as another instance

> For most case, we should be fine because we just check the workspace's ACL and workspace document is small; however if user want to access a specified saved object, we'll get the workspace of that saved object first then check that workspace's ACL to determine if you user has necessary permission. If user has permission, we'll read the object again to fetch the required fields. So permission check will add one more call. To optimize it, we can consider reading the docvalue of workspace field so the whole saved object won't be loaded. Will that work? If so, will need to update repository to add docvalue field as an optional parameter?

Yes! I think that will be perfect and a lot safer。

kavilla commented:
I see the cypress test failure. Would you like me to run it against some branch or fork of the FTrepo?

xluo-aws commented:
> Sounds good, do y'all have a board or a meta issue that is tracking things that will be come back to before finalizing ?

Here is the meta issue: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4944. I'm checking with PM if we need to support disable ACL flag because there is some concern from engineer side whether it should be disabled.

xluo-aws commented:
> I see the cypress test failure. Would you like me to run it against some branch or fork of the FTrepo?

The cypress test was successful when we submit this PR. It fails due to timeout, I don't have permission to rerun the test, could you help run the test again?

gaobinlong commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,250 @@
+}
+
+const addToPrincipals = (principals?: Principals, users?: string[], groups?: string[]) => {
+  if (!principals) {
```
I've changed that, thanks for your suggestion.

gaobinlong commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,250 @@
+  permissions: string[];
+}
+
+const addToPrincipals = (principals?: Principals, users?: string[], groups?: string[]) => {
```
That's a good point, I've changed that, thanks for your review.

gaobinlong commented:
> 
> Even if the user submits a request for new data or adds a new filter it is not refreshing it's saved objects data. Could it be possible for some admin of a workspace updated the actual information stored within `.kibana`. Which means dashboards and visualizations can be still populating with data to stuff they no longer have permissions until they refresh the page or navigate to a part of the application that update the "cached" saved objects. I don't believe it's a fully cache saved objects but the saved objects are not refreshed on every page.

If admin doesn't want someone to see the dashboards and visualizations, only modify the acl of the dashboards and visualizations is not enough, the backend data's permission should also be changed, if not users can still see the data by calling OpenSearch API, and if the backend data's permission is changed, then the new request for the data will be blocked,  so maybe it's not a big problem if users can still see the view page for a period of time before they refresh the page?

gaobinlong commented:
> > @kavilla, Rocky, the code that calls ACL functions will be submitted in another PR so it has to wait for this PR to be merged first. You can either reference https://github.com/ruanyl/OpenSearch-Dashboards/tree/workspace or we can create a draft PR if broken build is acceptable. What's your preference?
> 
> I think a draft on the usage will be helpful. Or a set a files to look at within that branch.
> 
> Thank you.

The caller of these ACL related functions are distributed in different PRs, but we can see them in our dev branch, here is the detail:

- validate if users/groups has permission: [validate()](https://github.com/ruanyl/OpenSearch-Dashboards/blob/29ccc203b591141aaf4486e361ec31acf1b93be4/src/plugins/workspace/server/permission_control/client.ts#L46)
- construct permissions object to store in `.kibana`: [convertToACL()](https://github.com/ruanyl/OpenSearch-Dashboards/blob/29ccc203b591141aaf4486e361ec31acf1b93be4/src/plugins/workspace/server/routes/index.ts#L52)
- convert from permissions object for display in UI: [convertFromACL()](https://github.com/ruanyl/OpenSearch-Dashboards/blob/29ccc203b591141aaf4486e361ec31acf1b93be4/src/plugins/workspace/server/routes/index.ts#L75)
- generate a query DSL to find all or some specified saved objects which users/groups have permission: [getQueryParams()](https://github.com/ruanyl/OpenSearch-Dashboards/blob/29ccc203b591141aaf4486e361ec31acf1b93be4/src/core/server/saved_objects/service/lib/search_dsl/query_params.ts#L290)

gaobinlong commented:
The code of consuming the `permissions` object in [repository](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5084/files#diff-01c073903639e29be8562ed0de79fae4084a953304672d1e6bdb3fd7fef06f11) are also added to this PR, which can make this PR more completer, unit tests for this change are also added,  please help to review, thanks!

xluo-aws commented:
@peternied , Hi, Peter, this is related to access control. Could you help take a look at it? Thanks!

peternied commented:
@xluo-aws @gaobinlong thanks for looking to include me; I won't be able to review this change.

zengyan-amazon commented on src/core/server/saved_objects/migrations/core/build_active_mappings.ts:
```diff
@@ -175,6 +186,15 @@ function defaultMapping(): IndexMapping {
+        properties: {
+          read: principals,
+          write: principals,
+          management: principals,
+          library_read: principals,
+          library_write: principals,
```
these permissions are specific to workspace object, shall we move it to workspaces?

Or in general can we have all ACL related settings in plugin instead of defining them in `saved_object` service?

gaobinlong replied:
Some functions such as `hasPermission`, `addPermission` can be moved to the workspace plugin, but the mapping change here, and the definition of the permissions object which is used by repository and saved_object_client cannot be moved to the workspace plugin.

zengyan-amazon commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,334 @@
+  users?: string[];
+  groups?: string[];
+}) => {
+  if (!!users) {
```
is it necessary to use double exclamation (!!) ?

gaobinlong replied:
No, it's not necessary, I'll change it, thanks!

zengyan-amazon commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,334 @@
+  return principals;
+};
+
+const checkPermission = (currentPrincipals: Principals | undefined, principals: Principals) => {
```
The naming of the function and its parameters are not intuitive, consider changing `currentPrincipals` to `allowedPrincipals`, and `principals` to `requesterPrincipal` 

btw, `principals` indicates it is a collection, while it is indeed one single identity. I think the use case is only check if a request identity has permission to have access to a saved object, so maybe not using principal type for requester?

gaobinlong replied:
Changing the name of the variables makes sense, thanks! For the principal type, I think its purpose is to make it consistent with the UI, the UI need to display the principal type: user or group, so when adding/removing/checking permissions, the caller need to use principal type explicitly.

zengyan-amazon commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,334 @@
+  );
+};
+
+export class ACL {
+  private permissions?: Permissions;
+  constructor(initialPermissions?: Permissions) {
+    this.permissions = initialPermissions || {};
+  }
+
```
ACL and Permissions seems the the same thing, shall we consider to move the functions to Permissions to reduce the complexity?

gaobinlong replied:
You mean change the class name `ACL` to `Permissions`? `Permissions` is a permissionType->Principals map, just a object which has the same structure with the field in .kibana index, ACL contains the Permissions object and some fundamental functions, I think they are not the same.

zengyan-amazon replied:
by simply looking at the code, the `ACL` class has only one field, which is `permissions?: Permissions` , so all operations `ACL` can provide is based on the `permissions` right? so just don't feel we don't need 2 levels of classes here

gaobinlong replied:
`export type Permissions = Record<string, Principals>;`

`Permissions` is a type, not class, and I think we need two level of definitions here is because `Permissions` is referenced by saved_object_client and repository, the relationship between `ACL` and `Permissions` just like the relationship between [SavedObjectsSerializer](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/9e3e3a7ae8e950c5e4137ebf5dafd8f093099f0b/src/core/server/saved_objects/serialization/serializer.ts#L44) and `SavedObjectsRawDoc`.

zengyan-amazon replied:
okay, the name `ACL` doesn't sounds right here, maybe `PermissionEvaluator` makes more sense

zengyan-amazon replied:
And should this go here in the saved object service? or workspace service?

gaobinlong replied:
Since the core module in OSD cannot depend on the code of plugin, but we have to add the field `permissions` to `SavedObjectsRepository`, `saved_objects_client` and `saved_objects` in the core module, so we cannot move the ACL class to the workspace plugin.

zengyan-amazon commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,334 @@
+      this.permissions = {};
+    }
+
+    for (const permissionType of permissionTypes) {
```
in the case if a user has `write` permission to an object, and we want to remove the (`read`) access to the object, how do we ensure the user principal is removed from the `write` list? is it controlled by the caller of this `removePermission` function?

gaobinlong replied:
It's controlled by the caller, the caller needs to call `removePermission` and `addPermission` functions multiple times to construct a new `Permissions` object.

zengyan-amazon commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,334 @@
+   * @public
+   * @static
+   */
+  public static generateGetPermittedSavedObjectsQueryDSL(
```
where is this function being used?

gaobinlong replied:
You can see the caller in our dev branch: [getQueryParams()](https://github.com/ruanyl/OpenSearch-Dashboards/blob/29ccc203b591141aaf4486e361ec31acf1b93be4/src/core/server/saved_objects/service/lib/search_dsl/query_params.ts#L290), it's used when generating a query DSL to find all or some specified saved objects which users/groups have permissions.

zengyan-amazon reviewed (commented):
I have a few high level questions:
* Given ACL is a breaking experience, we discussed allowing admin to enable/disable ACL using a yml config, but it seems this implementation doesn't support this feature flag
* I actually think of this saved object access control to be implemented as something like a plugin, instead of instrument SavedObject service deeply, is there any technical blocker for doing this in plugin?

gaobinlong commented:
> I have a few high level questions:
> 
> * Given ACL is a breaking experience, we discussed allowing admin to enable/disable ACL using a yml config, but it seems this implementation doesn't support this feature flag
> * I actually think of this saved object access control to be implemented as something like a plugin, instead of instrument SavedObject service deeply, is there any technical blocker for doing this in plugin?

Yeah, there will be another PR to implement the feature flag of ACL, not in this one. And in our current implementation, saved object access control is not a plugin, we may do some refactor for the current code, so it's in our future plan.

Flyingliuhub commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,337 @@
+   *   .addPermission(['write', 'management'], {
+   *     groups: ['group1'],
+   *   })
+   *   .getPermissions();;
```
Extra semicolon for `.getPermissions();;`

gaobinlong replied:
Thanks for catching that, I've removed it.

Flyingliuhub commented on src/core/server/saved_objects/permission_control/acl.ts:
```diff
@@ -0,0 +1,337 @@
+   *   .removePermission(['write', 'management'], {
+   *     groups: ['group1'],
+   *   })
+   *   .getPermissions();;
```
Extra semicolon for `.getPermissions();;`

Flyingliuhub commented on CHANGELOG.md:
```diff
@@ -52,6 +52,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Replace OuiSelect component with OuiSuperSelect in data-source plugin ([#5315](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5315))
 - [Workspace] Add core workspace service module to enable the implementation of workspace features within OSD plugins ([#5092](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5092))
 - [Workspace] Setup workspace skeleton and implement basic CRUD API ([#5075](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5075))
+- [Workspace] Add ACL related functions ([#5084](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5084/))
```
Is it possible to create a short video and add it to this PR, so everyone can visually see the permission control behavior?

gaobinlong replied:
I'm glad to do that, but for this PR, it only works when the caller triggers the functions call, and the code of the caller will be in the following PR, so it's hard to tigger these basic functions in this PR by UI now. We will create a video for permission control in the following PR later.

ashwin-pc commented:
@gaobinlong can you add details to the "Testing the changes" section of the PR description? How can i pull down and test this PR?

gaobinlong commented:
> @gaobinlong can you add details to the "Testing the changes" section of the PR description? How can i pull down and test this PR?

Thanks for your comment, I know it's important that the reviewer know how to test these change by UI, but for this PR, it contains only some basic functions which will be called by the saved objects APIs, the code of the caller will be in the following PRs, so we'll elaborate more about how to test these change by UI in that PR later.

Flyingliuhub commented on .github/workflows/build_and_test_workflow.yml:
```diff
@@ -6,7 +6,7 @@ name: Build and test
 on:
   push:
-    branches: ['**', '!backport/**']
+    branches: ['main', '[0-9].x', '[0-9].[0=9]+'] # Run the functional test on push for only release branches
```
It seems like that your commits includes a lot changes which already merged recently, can you rebase your branch, and make sure this PR only have your changes? Thanks

gaobinlong replied:
I've forced push the branch to remove the duplicated commits, thanks!

gaobinlong commented:
> I actually think of this saved object access control to be implemented as something like a plugin, instead of instrument SavedObject service deeply, is there any technical blocker for doing this in plugin?

Some change should happen in the file of  core module like `savedObjetcsRepository`, `saved_object_client` and `saved_objects`, but the code of the core module cannot depend on plugin, so it's hard to make access control as a plugin or move it to plugin.

gaobinlong commented:
Hi @zengyan-amazon @Flyingliuhub @ashwin-pc @SuZhou-Joe @ruanyl , according to our discussion, I've added a independent feature flag `savedObjects.permission.enabled` which can control whether permission control for saved objects is enabled or not, and then we check the flag in savedObjectsService, if true, we change the mapping of the OSD system index, if false, nothing changes. Please help to take a look at the latest code change, thanks!

Flyingliuhub commented on src/core/server/saved_objects/migrations/opensearch_dashboards/opensearch_dashboards_migrator.test.ts:
```diff
@@ -76,6 +77,12 @@ describe('OpenSearchDashboardsMigrator', () => {
       expect(mappings).toMatchSnapshot();
     });
+
+    it('permissions field exists in the mappings when the feature is enabled', () => {
```
do we cover the tests that permissons field not exists in the mapping when the feature is disabled? Thanks

gaobinlong replied:
Yes, the mocked value of the feature flag is false for all other tests, and they will check if the generated mappings match the snapshot which doesn't contain the `permissions` field: ` expect(mappings).toMatchSnapshot();`.

gaobinlong commented on src/core/server/saved_objects/saved_objects_config.ts:
```diff
@@ -49,6 +49,9 @@ export const savedObjectsConfig = {
     maxImportPayloadBytes: schema.byteSize({ defaultValue: 26214400 }),
     maxImportExportSize: schema.byteSize({ defaultValue: 10000 }),
+    permission: schema.object({
+      enabled: schema.boolean({ defaultValue: false }),
```
Thanks, I've done that, please help to take a look.

xluo-aws commented on config/opensearch_dashboards.yml:
```diff
@@ -285,4 +285,8 @@
+# vis_augmenter.pluginAugmentationEnabled: true
+
+# Set the value to true to enable permission control for saved objects
+# Permission control depends on OpenSearch Dashboards has authentication enabled, set it to false if no authentication is configured
```
What's the configuration for OSD authentication flag?

Hailong-am replied:
should be security plugin installed or not

gaobinlong replied:
Yeah, there's no authentication flag, I've modified the comment to make it more clear.

ruanyl commented on src/core/server/saved_objects/migrations/core/build_active_mappings.ts:
```diff
@@ -47,11 +49,36 @@ import {
+        properties: {
+          read: principals,
+          write: principals,
+          management: principals,
```
Discussed this with @zengyan-amazon, the `management` permission is essentially the same as `write` permission which allows the principals to update/delete the workspace object itself. In that sense, a user with `write` and `library_write` should be sufficient to make the user a workspace admin.

In the current implementation of workspace creation and workspace permission validation, we actually no longer rely on `management`. So we don't need to have it in the index mapping.

xluo-aws replied:
make sense

gaobinlong replied:
I've removed the management permission type yet, please help to take a look, thanks!

ruanyl reviewed (approved):
Excellent job! Appreciate your dedication to refining the code according to the comments and discussions. It looks good to me now. 👍
